File: v4/rust/iprange-livedb/src/publication/resolver_result.rs
//! Structured result construction after resolver classification.

use crate::cancellation::CancellationToken;
use crate::error::ErrorCode;

use super::super::result::{
    AccessPolicy, CleanupArtifacts, DestinationContent, FinalState, LaterCanonical,
    PublicationResult, PublicationStatus, Seed,
};
use super::{
    cleanup, file_inspection, reservation_identity, verify_later, Destination, Header, Inspected,
    Location, Problem,
};

pub(super) struct DesiredContext<'a> {
    pub(super) destination: &'a Destination,
    pub(super) header: Header,
    pub(super) reservation: Option<&'a Inspected>,
    pub(super) later: Option<&'a Inspected>,
    pub(super) main: &'a file_inspection::Inspected,
}

pub(super) fn record_cancellation(
    mut result: PublicationResult,
    cancellation: &CancellationToken,
) -> PublicationResult {
    let cleanup_cause = result.cleanup.get(0).map(|artifact| artifact.error.clone());
    if cancellation.is_cancelled() && (result.cause.is_none() || result.cause == cleanup_cause) {
        result.cause = Some(Problem::new(
            ErrorCode::Cancelled,
            None,
            "publication resolution was cancelled after mutation",
        ));
    }
    result
}

pub(super) fn desired_result(
    seed: Seed,
    summary: cleanup::Summary,
    context: DesiredContext<'_>,
) -> PublicationResult {
    let DesiredContext {
        destination,
        header,
        reservation,
        later,
        main,
    } = context;
    let verification = main
        .verify(destination)
        .and_then(|()| verify_later(later, destination));
    let cause = verification
        .as_ref()
        .err()
        .cloned()
        .or_else(|| first_problem(&summary.artifacts));
    seed.result_with_housekeeping(
        desired_state(
            header,
            main,
            &summary,
            reservation,
            later,
            verification.is_ok(),
        ),
        summary.artifacts,
        summary.housekeeping,
        summary.visible_housekeeping,
        cause,
    )
    .with_later(later)
}

pub(super) fn desired_problem(
    seed: Seed,
    header: Header,
    summary: cleanup::Summary,
    problem: Problem,
) -> PublicationResult {
    seed.result_with_housekeeping(
        FinalState {
            reservation_identity: reservation_identity(header),
            main_namespace_may_have_been_attempted: true,
            publication: PublicationStatus::OutcomeUnknown,
            destination_content: DestinationContent::Unclassified,
            main_access_policy: AccessPolicy::Unclassified,
            coordination_access_policy: AccessPolicy::Unclassified,
        },
        summary.artifacts,
        summary.housekeeping,
        summary.visible_housekeeping,
        Some(problem),
    )
}

pub(super) fn published_output_result(
    seed: Seed,
    summary: cleanup::Summary,
    problem: Problem,
    context: DesiredContext<'_>,
) -> PublicationResult {
    let DesiredContext {
        destination,
        header,
        reservation,
        later,
        main,
    } = context;
    let verified = main
        .verify(destination)
        .and_then(|()| verify_later(later, destination))
        .is_ok();
    seed.result_with_housekeeping(
        desired_state(header, main, &summary, reservation, later, verified),
        summary.artifacts,
        summary.housekeeping,
        summary.visible_housekeeping,
        Some(problem),
    )
    .with_later(later)
}

fn desired_state(
    header: Header,
    main: &file_inspection::Inspected,
    summary: &cleanup::Summary,
    reservation: Option<&Inspected>,
    later: Option<&Inspected>,
    verified: bool,
) -> FinalState {
    FinalState {
        reservation_identity: reservation_identity(header),
        main_namespace_may_have_been_attempted: true,
        publication: if verified {
            PublicationStatus::Published
        } else {
            PublicationStatus::OutcomeUnknown
        },
        destination_content: if verified {
            DestinationContent::Desired
        } else {
            DestinationContent::Unclassified
        },
        main_access_policy: if verified {
            main.access
        } else {
            AccessPolicy::Unclassified
        },
        coordination_access_policy: if verified {
            coordination_access(summary, reservation, later)
        } else {
            AccessPolicy::Unclassified
        },
    }
}

pub(super) fn coordination_access(
    summary: &cleanup::Summary,
    reservation: Option<&Inspected>,
    later: Option<&Inspected>,
) -> AccessPolicy {
    if summary.coordination_absent {
        AccessPolicy::Absent
    } else {
        later
            .or_else(|| {
                reservation.filter(|reservation| reservation.location == Location::Canonical)
            })
            .map_or(AccessPolicy::Unclassified, |reservation| reservation.access)
    }
}

pub(super) fn first_problem(cleanup: &CleanupArtifacts) -> Option<Problem> {
    cleanup.get(0).map(|artifact| artifact.error.clone())
}

pub(super) trait LaterResult {
    fn with_later(self, later: Option<&Inspected>) -> Self;
}

impl LaterResult for PublicationResult {
    fn with_later(mut self, later: Option<&Inspected>) -> Self {
        if later.is_some_and(|reservation| reservation.location == Location::Canonical) {
            self.later_canonical = LaterCanonical::ReservationOrTransition;
        }
        self
    }
}

File: v4/go/internal/publication/resolver_result.go
//go:build !windows

// Structured result construction after resolver classification (Rust
// publication/resolver_result.rs).

package publication

import (
	"errors"

	"github.com/firehol/iprange/v4/go/internal/format"
)

// desiredContext is the shared fact inventory of one desired-main
// resolution (Rust DesiredContext).
type desiredContext struct {
	destination *destination
	header      reservationHeader
	reservation *inspectedReservation
	later       *inspectedReservation
	main        *inspectedOutput
}

// recordCancellation folds one post-mutation cancellation into the
// result cause (Rust record_cancellation): when the caller cancelled
// during the cleanup and the result carries no cause or only the
// cleanup cause, the fixed cancellation class replaces it.
func recordCancellation(result PublicationResult, check func() error) PublicationResult {
	if err := checkCancellation(check); err == nil {
		return result
	}
	cleanupCause := firstProblem(&result.Cleanup)
	if result.Cause == nil || problemEqual(result.Cause, cleanupCause) {
		result.Cause = problem(format.CodeCancelled, "publication resolution was cancelled after mutation")
	}
	return result
}

// problemEqual compares two problems by code and detail (Rust
// Problem PartialEq on the cleanup clone; fresh constructions of one
// fixed problem are distinct pointers but equal values).
func problemEqual(a, b error) bool {
	if a == nil && b == nil {
		return true
	}
	if a == nil || b == nil {
		return false
	}
	var fa, fb *format.Error
	if errors.As(a, &fa) && errors.As(b, &fb) {
		return fa.Code == fb.Code && fa.Detail == fb.Detail
	}
	return a == b
}

// desiredResult builds the exact published result of one desired main
// (Rust desired_result).
func desiredResult(s seed, summary cleanupSummary, context desiredContext) PublicationResult {
	verification := context.main.verify(context.destination)
	if verification == nil {
		verification = verifyLater(context.later, context.destination)
	}
	cause := verification
	if cause == nil {
		cause = firstProblem(&summary.artifacts)
	}
	result := s.resultWithHousekeeping(
		desiredState(context.header, context.main, summary, context.reservation, context.later, verification == nil),
		summary.artifacts, summary.housekeeping, summary.visibleHousekeeping, cause,
	)
	return withLater(result, context.later)
}

// desiredProblem builds the outcome-unknown result of one failed
// desired-main cleanup (Rust desired_problem).
func desiredProblem(s seed, header reservationHeader, summary cleanupSummary, problem error) PublicationResult {
	return s.resultWithHousekeeping(finalState{
		reservationIdentity:               reservationIdentityOf(header),
		mainNamespaceMayHaveBeenAttempted: true,
		publication:                       PublicationOutcomeUnknown,
		destinationContent:                DestinationContentUnclassified,
		mainAccessPolicy:                  AccessPolicyUnclassified,
		coordinationAccessPolicy:          AccessPolicyUnclassified,
	}, summary.artifacts, summary.housekeeping, summary.visibleHousekeeping, problem)
}

// publishedOutputResult builds the published result with the exact
// private-output failure cause (Rust published_output_result).
func publishedOutputResult(s seed, summary cleanupSummary, problem error, context desiredContext) PublicationResult {
	verified := context.main.verify(context.destination)
	if verified == nil {
		verified = verifyLater(context.later, context.destination)
	}
	result := s.resultWithHousekeeping(
		desiredState(context.header, context.main, summary, context.reservation, context.later, verified == nil),
		summary.artifacts, summary.housekeeping, summary.visibleHousekeeping, problem,
	)
	return withLater(result, context.later)
}

// desiredState builds the final state of one desired main (Rust
// desired_state).
func desiredState(header reservationHeader, main *inspectedOutput, summary cleanupSummary, reservation *inspectedReservation, later *inspectedReservation, verified bool) finalState {
	state := finalState{
		reservationIdentity:               reservationIdentityOf(header),
		mainNamespaceMayHaveBeenAttempted: true,
		publication:                       PublicationOutcomeUnknown,
		destinationContent:                DestinationContentUnclassified,
		mainAccessPolicy:                  AccessPolicyUnclassified,
		coordinationAccessPolicy:          AccessPolicyUnclassified,
	}
	if verified {
		state.publication = PublicationPublished
		state.destinationContent = DestinationContentDesired
		state.mainAccessPolicy = main.access
		state.coordinationAccessPolicy = coordinationAccess(summary, reservation, later)
	}
	return state
}

// coordinationAccess derives the coordination access policy of one
// post-cleanup state (Rust coordination_access).
func coordinationAccess(summary cleanupSummary, reservation *inspectedReservation, later *inspectedReservation) AccessPolicy {
	if summary.coordinationAbsent {
		return AccessPolicyAbsent
	}
	owner := later
	if owner == nil && reservation != nil && reservation.location == reservationLocationCanonical {
		owner = reservation
	}
	if owner == nil {
		return AccessPolicyUnclassified
	}
	return owner.access
}

// firstProblem reports the first cleanup artifact problem (Rust
// first_problem).
func firstProblem(cleanup *CleanupArtifacts) error {
	if cleanup == nil || cleanup.Empty() {
		return nil
	}
	return cleanup.At(0).Error
}

// withLater records the later-canonical observation (Rust
// LaterResult::with_later).
func withLater(result PublicationResult, later *inspectedReservation) PublicationResult {
	if later != nil && later.location == reservationLocationCanonical {
		result.LaterCanonical = LaterCanonicalReservationOrTransition
	}
	return result
}
